File: examples/primitives/hash/validation.ts
/**
 * Hash Validation Example
 *
 * Demonstrates:
 * - Validating hex strings before parsing
 * - Type guards and runtime type checking
 * - Safe parsing patterns
 * - Assertion-based validation
 * - Form input validation
 */

import { Hash } from "../../../src/primitives/Hash/index.js";

const testCases = [
	{
		input: "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
		expected: true,
		reason: "Valid with 0x prefix",
	},
	{
		input: "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
		expected: true,
		reason: "Valid without 0x prefix",
	},
	{ input: "0x1234", expected: false, reason: "Too short (4 chars, need 64)" },
	{
		input:
			"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdefff",
		expected: false,
		reason: "Too long (66 chars)",
	},
	{
		input: "0xGGGG567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
		expected: false,
		reason: "Invalid hex chars",
	},
	{
		input: "0x1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF",
		expected: true,
		reason: "Valid uppercase",
	},
	{ input: "", expected: false, reason: "Empty string" },
	{ input: "0x", expected: false, reason: "Only prefix" },
];
testCases.forEach(({ input, expected, reason }) => {
	const isValid = Hash.isValidHex(input);
	const status = isValid === expected ? "✓" : "✗";
	const displayInput =
		input.length > 20 ? `${input.slice(0, 10)}...${input.slice(-6)}` : input;
});

// Pattern 1: Return null on invalid
function parseHashOrNull(input: string): Hash | null {
	if (!Hash.isValidHex(input)) {
		return null;
	}
	try {
		return Hash.fromHex(input);
	} catch (error) {
		console.error(
			`Parse error: ${error instanceof Error ? error.message : "Unknown error"}`,
		);
		return null;
	}
}

// Pattern 2: Throw on invalid
function parseHashOrThrow(input: string): Hash {
	if (!Hash.isValidHex(input)) {
		throw new Error(`Invalid hash format: ${input}`);
	}
	return Hash.fromHex(input);
}

// Pattern 3: Return result object
interface ParseResult {
	success: boolean;
	hash?: Hash;
	error?: string;
}

function parseHashResult(input: string): ParseResult {
	if (!Hash.isValidHex(input)) {
		return { success: false, error: "Invalid hex format" };
	}
	try {
		return { success: true, hash: Hash.fromHex(input) };
	} catch (error) {
		return {
			success: false,
			error: error instanceof Error ? error.message : "Unknown error",
		};
	}
}

// Test patterns
const validInput =
	"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef";
const invalidInput = "0x1234";
const result1 = parseHashOrNull(validInput);
const result2 = parseHashOrNull(invalidInput);
try {
	const result3 = parseHashOrThrow(validInput);
} catch (error) {
}

try {
	const result4 = parseHashOrThrow(invalidInput);
} catch (error) {
}
const result5 = parseHashResult(validInput);
const result6 = parseHashResult(invalidInput);

function processUnknownValue(value: unknown, label: string) {
	if (Hash.isHash(value)) {
	} else if (typeof value === "string") {
	} else if (value instanceof Uint8Array) {
	} else {
	}
}

const hash = Hash.random();
const bytes32 = new Uint8Array(32);
const bytes20 = new Uint8Array(20);
const hexString =
	"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef";
const number = 123;
processUnknownValue(hash, "Hash instance");
processUnknownValue(bytes32, "32 bytes");
processUnknownValue(bytes20, "20 bytes");
processUnknownValue(hexString, "Hex string");
processUnknownValue(number, "Number");

const mixedValues: unknown[] = [
	Hash.random(),
	new Uint8Array(32),
	new Uint8Array(20),
	"not a hash",
	Hash.keccak256String("hello"),
	123,
	null,
];

const validHashes = mixedValues.filter(Hash.isHash);
validHashes.forEach((hash, i) => {
});

function processHash(value: unknown): string {
	// Assert will throw if not a hash
	Hash.assert(value, "Expected hash value");

	// TypeScript knows value is Hash after assertion
	return Hash.toHex(value);
}

try {
	const validHash = Hash.random();
	const result = processHash(validHash);
} catch (error) {
}

try {
	const invalidValue = "not a hash";
	const result = processHash(invalidValue);
} catch (error) {
}

interface ValidationError {
	field: string;
	message: string;
}

function validateHashInput(input: string): ValidationError[] {
	const errors: ValidationError[] = [];

	// Required check
	if (!input || input.trim() === "") {
		errors.push({ field: "hash", message: "Hash is required" });
		return errors;
	}

	// Prefix check
	if (!input.startsWith("0x")) {
		errors.push({ field: "hash", message: "Hash must start with 0x prefix" });
	}

	// Length check
	const normalized = input.slice(2);
	if (normalized.length !== 64) {
		errors.push({
			field: "hash",
			message: `Hash must be 64 hex characters, got ${normalized.length}`,
		});
	}

	// Character check
	if (!/^[0-9a-fA-F]+$/.test(normalized)) {
		errors.push({ field: "hash", message: "Hash contains invalid characters" });
	}

	// Or use isValidHex
	if (errors.length === 0 && !Hash.isValidHex(input)) {
		errors.push({ field: "hash", message: "Invalid hash format" });
	}

	return errors;
}

const formInputs = [
	"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", // Valid
	"", // Empty
	"1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", // No prefix
	"0x1234", // Too short
	"0xGGGG567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", // Invalid chars
];
formInputs.forEach((input) => {
	const errors = validateHashInput(input);
	const display =
		input.length > 20 ? `${input.slice(0, 10)}...` : input || "(empty)";

	if (errors.length === 0) {
	} else {
		errors.forEach((err) => );
	}
});

interface GetLogsParams {
	fromBlock: number;
	toBlock: number;
	topics: (string | null)[];
}

function validateGetLogsParams(params: GetLogsParams): void {
	// Validate block range
	if (params.fromBlock < 0) {
		throw new Error("fromBlock must be non-negative");
	}
	if (params.toBlock < params.fromBlock) {
		throw new Error("toBlock must be >= fromBlock");
	}

	// Validate topics
	for (let i = 0; i < params.topics.length; i++) {
		const topic = params.topics[i];
		if (topic !== null && !Hash.isValidHex(topic)) {
			throw new Error(`Invalid topic hash at index ${i}: ${topic}`);
		}
	}
}

// Valid params
try {
	const validParams: GetLogsParams = {
		fromBlock: 1000,
		toBlock: 2000,
		topics: [
			"0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
			null,
			null,
		],
	};
	validateGetLogsParams(validParams);
} catch (error) {
}

// Invalid params
try {
	const invalidParams: GetLogsParams = {
		fromBlock: 1000,
		toBlock: 2000,
		topics: ["0x1234"], // Too short
	};
	validateGetLogsParams(invalidParams);
} catch (error) {
}

const hashes = [
	Hash.fromBytes(new Uint8Array(32)), // Zero
	Hash.keccak256String("hello"), // Non-zero
	Hash.ZERO, // Zero constant
	Hash.random(), // Non-zero
];
hashes.forEach((hash, i) => {
	const isZero = hash.isZero();
	const status = isZero ? "ZERO" : "NON-ZERO";
});

function validateHashBatch(inputs: string[]): {
	valid: Hash[];
	invalid: string[];
} {
	const valid: Hash[] = [];
	const invalid: string[] = [];

	for (const input of inputs) {
		if (Hash.isValidHex(input)) {
			try {
				valid.push(Hash.fromHex(input));
			} catch {
				invalid.push(input);
			}
		} else {
			invalid.push(input);
		}
	}

	return { valid, invalid };
}

const batchInputs = [
	"0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
	"0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890",
	"0x1234", // Invalid
	"0xGGGG567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", // Invalid
	"0xfedcba0987654321fedcba0987654321fedcba0987654321fedcba0987654321",
];

const batchResult = validateHashBatch(batchInputs);
batchResult.invalid.forEach((input) => {
	const display = input.length > 20 ? `${input.slice(0, 10)}...` : input;
});

File: packages/voltaire-zig/examples/primitives/hash/validation.zig
const std = @import("std");
const primitives = @import("primitives");
const Hash = primitives.hash.Hash;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const a = gpa.allocator();

    std.debug.print("\n=== Hash Validation Example ===\n\n", .{});

    // ============================================================
    // 1. Hex String Validation
    // ============================================================

    std.debug.print("1. Hex String Validation\n\n", .{});

    const TestCase = struct {
        input: []const u8,
        expected: bool,
        reason: []const u8,
    };

    const test_cases = [_]TestCase{
        .{ .input = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", .expected = true, .reason = "Valid with 0x prefix" },
        .{ .input = "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", .expected = true, .reason = "Valid without 0x prefix" },
        .{ .input = "0x1234", .expected = false, .reason = "Too short" },
        .{ .input = "0xGGGG567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", .expected = false, .reason = "Invalid hex chars" },
        .{ .input = "0x1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF", .expected = true, .reason = "Valid uppercase" },
        .{ .input = "", .expected = false, .reason = "Empty string" },
        .{ .input = "0x", .expected = false, .reason = "Only prefix" },
    };

    std.debug.print("Test cases:\n", .{});
    for (test_cases) |case| {
        const is_valid = Hash.isValid(case.input);
        const status: u8 = if (is_valid == case.expected) '✓' else '✗';
        const display = if (case.input.len > 20) blk: {
            var buf: [20]u8 = undefined;
            @memcpy(buf[0..10], case.input[0..10]);
            @memcpy(buf[10..17], "...");
            break :blk buf[0..17];
        } else case.input;
        std.debug.print("  {c} {s:<20} → {} ({s})\n", .{ status, display, is_valid, case.reason });
    }
    std.debug.print("\n", .{});

    // ============================================================
    // 2. Safe Parsing Patterns
    // ============================================================

    std.debug.print("2. Safe Parsing Patterns\n\n", .{});

    // Pattern 1: Return optional (null on error)
    const valid_input = "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef";
    const invalid_input = "0x1234";

    std.debug.print("Pattern 1 (optional):\n", .{});

    // Valid
    if (Hash.isValid(valid_input)) {
        const result1 = Hash.fromHex(valid_input) catch null;
        if (result1) |hash| {
            const hex = Hash.toHex(hash);
            std.debug.print("  Valid:   0x{s}...{s}\n", .{ hex[2..8], hex[hex.len - 4 ..] });
        } else {
            std.debug.print("  Valid:   null\n", .{});
        }
    } else {
        std.debug.print("  Valid:   invalid format\n", .{});
    }

    // Invalid
    if (Hash.isValid(invalid_input)) {
        const result2 = Hash.fromHex(invalid_input) catch null;
        if (result2) |hash| {
            const hex = Hash.toHex(hash);
            std.debug.print("  Invalid: 0x{s}...{s}\n", .{ hex[2..8], hex[hex.len - 4 ..] });
        } else {
            std.debug.print("  Invalid: null\n", .{});
        }
    } else {
        std.debug.print("  Invalid: invalid format\n", .{});
    }

    std.debug.print("\nPattern 2 (error return):\n", .{});

    // Valid
    const result3 = Hash.fromHex(valid_input) catch |err| {
        std.debug.print("  Error: {}\n", .{err});
        return err;
    };
    const hex3 = Hash.toHex(result3);
    std.debug.print("  Valid:   0x{s}...{s}\n", .{ hex3[2..8], hex3[hex3.len - 4 ..] });

    // Invalid
    _ = Hash.fromHex(invalid_input) catch |err| {
        std.debug.print("  Invalid: Error: {}\n", .{err});
    };

    std.debug.print("\n", .{});

    // ============================================================
    // 3. Type Safety (Compile-Time)
    // ============================================================

    std.debug.print("3. Type Safety\n\n", .{});

    // In Zig, Hash is [32]u8 - type checked at compile time
    const hash = Hash.random();
    const bytes32: [32]u8 = hash; // Valid
    const hex_typed = Hash.toHex(bytes32);
    std.debug.print("  Hash instance: 0x{s}...{s}\n", .{ hex_typed[2..8], hex_typed[hex_typed.len - 4 ..] });

    // Wrong length would cause compile error
    // var bytes20: [20]u8 = undefined;
    // const invalid: Hash = bytes20; // Compile error!

    std.debug.print("  (Zig provides compile-time type safety)\n\n", .{});

    // ============================================================
    // 4. Filtering Collections
    // ============================================================

    std.debug.print("4. Filtering Collections\n\n", .{});

    var valid_hashes = std.ArrayList(Hash).init(a);
    defer valid_hashes.deinit();

    // Simulate mixed values - in Zig we'd use tagged union
    const hash1 = Hash.random();
    const hash2 = try Hash.keccak256String("hello");

    try valid_hashes.append(hash1);
    try valid_hashes.append(hash2);

    std.debug.print("Valid hashes: {}\n", .{valid_hashes.items.len});
    std.debug.print("Filtered hashes:\n", .{});
    for (valid_hashes.items, 0..) |h, i| {
        const hex = Hash.toHex(h);
        std.debug.print("  {}. 0x{s}...{s}\n", .{ i + 1, hex[2..8], hex[hex.len - 4 ..] });
    }
    std.debug.print("\n", .{});

    // ============================================================
    // 5. API Parameter Validation
    // ============================================================

    std.debug.print("5. API Parameter Validation\n\n", .{});

    const GetLogsParams = struct {
        from_block: u64,
        to_block: u64,
        topics: []const ?[]const u8,
    };

    const validateGetLogsParams = struct {
        fn validate(params: GetLogsParams) !void {
            if (params.to_block < params.from_block) {
                return error.InvalidBlockRange;
            }

            for (params.topics, 0..) |topic, i| {
                if (topic) |t| {
                    if (!Hash.isValid(t)) {
                        std.debug.print("Invalid topic at index {}\n", .{i});
                        return error.InvalidTopic;
                    }
                }
            }
        }
    }.validate;

    // Valid params
    const valid_topics = [_]?[]const u8{
        "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef",
        null,
        null,
    };
    const valid_params = GetLogsParams{
        .from_block = 1000,
        .to_block = 2000,
        .topics = &valid_topics,
    };

    validateGetLogsParams(valid_params) catch |err| {
        std.debug.print("  ✗ Error: {}\n", .{err});
    };
    std.debug.print("  ✓ Valid params accepted\n", .{});

    // Invalid params
    const invalid_topics = [_]?[]const u8{"0x1234"}; // Too short
    const invalid_params = GetLogsParams{
        .from_block = 1000,
        .to_block = 2000,
        .topics = &invalid_topics,
    };

    validateGetLogsParams(invalid_params) catch |err| {
        std.debug.print("  ✗ Error: {}\n", .{err});
    };
    std.debug.print("\n", .{});

    // ============================================================
    // 6. Zero Hash Validation
    // ============================================================

    std.debug.print("6. Zero Hash Validation\n\n", .{});

    const hashes = [_]Hash{
        Hash.fromBytes(&([_]u8{0} ** 32)), // Zero
        try Hash.keccak256String("hello"), // Non-zero
        Hash.ZERO, // Zero constant
        Hash.random(), // Non-zero
    };

    std.debug.print("Zero hash checks:\n", .{});
    for (hashes, 0..) |h, i| {
        const is_zero = Hash.isZero(h);
        const status = if (is_zero) "ZERO" else "NON-ZERO";
        const hex = Hash.toHex(h);
        std.debug.print("  {}. 0x{s}...{s} → {s}\n", .{ i + 1, hex[2..8], hex[hex.len - 4 ..], status });
    }
    std.debug.print("\n", .{});

    // ============================================================
    // 7. Batch Validation
    // ============================================================

    std.debug.print("7. Batch Validation\n\n", .{});

    const batch_inputs = [_][]const u8{
        "0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef",
        "0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890",
        "0x1234", // Invalid
        "0xGGGG567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef", // Invalid
        "0xfedcba0987654321fedcba0987654321fedcba0987654321fedcba0987654321",
    };

    var valid_list = std.ArrayList(Hash).init(a);
    defer valid_list.deinit();
    var invalid_list = std.ArrayList([]const u8).init(a);
    defer invalid_list.deinit();

    for (batch_inputs) |input| {
        if (Hash.isValid(input)) {
            if (Hash.fromHex(input)) |hash| {
                try valid_list.append(hash);
            } else |_| {
                try invalid_list.append(input);
            }
        } else {
            try invalid_list.append(input);
        }
    }

    std.debug.print("Total inputs: {}\n", .{batch_inputs.len});
    std.debug.print("Valid: {}\n", .{valid_list.items.len});
    std.debug.print("Invalid: {}\n", .{invalid_list.items.len});
    std.debug.print("\nInvalid inputs:\n", .{});
    for (invalid_list.items) |input| {
        const display = if (input.len > 20) blk: {
            var buf: [17]u8 = undefined;
            @memcpy(buf[0..10], input[0..10]);
            @memcpy(buf[10..], "...");
            break :blk buf[0..];
        } else input;
        std.debug.print("  - {s}\n", .{display});
    }
    std.debug.print("\n", .{});

    std.debug.print("=== Example Complete ===\n\n", .{});
}
